modules/apps/message-boards/message-boards-service/src/main/java/com/liferay/message/boards/internal/util/MBMessageUtil.java

```diff
@@ -17,19 +17,195 @@
 import com.liferay.message.boards.model.MBMessage;
 import com.liferay.message.boards.service.persistence.MBMessageFinder;
 import com.liferay.message.boards.service.persistence.MBMessagePersistence;
+import com.liferay.petra.string.StringBundler;
+import com.liferay.petra.string.StringPool;
 import com.liferay.portal.kernel.dao.orm.QueryDefinition;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.parsers.bbcode.BBCodeTranslatorUtil;
+import com.liferay.portal.kernel.service.ServiceContext;
+import com.liferay.portal.kernel.theme.ThemeDisplay;
 import com.liferay.portal.kernel.util.ListUtil;
 import com.liferay.portal.kernel.util.OrderByComparator;
+import com.liferay.portal.kernel.util.Validator;
+import com.liferay.portal.kernel.util.WebKeys;
 import com.liferay.portal.kernel.workflow.WorkflowConstants;
 
 import java.util.Comparator;
 import java.util.List;
 
+import javax.servlet.http.HttpServletRequest;
+
 /**
  * @author Preston Crary
  */
 public class MBMessageUtil {
 
+	public static String getAncestryMessages(
+		Boolean htmlFormat, String quoteMark, List<MBMessage> thread,
+		ServiceContext serviceContext) {
+
+		String messageParentBody = StringPool.BLANK;
+
+		if (ListUtil.isEmpty(thread)) {
+			return messageParentBody;
+		}
+
+		String end = StringPool.BLANK;
+		String endElement = StringPool.BLANK;
+		StringBuilder endElements = new StringBuilder(StringPool.BLANK);
+		String endMessage = StringPool.BLANK;
+		String startMessage = StringPool.BLANK;
+		String startMessageBody = StringPool.BLANK;
+		String startMessageThread = StringPool.BLANK;
+		String startUser = StringPool.BLANK;
+
+		if (htmlFormat) {
+			end = "</div>";
+			endElement = "</ul>";
+			endMessage = "</li>";
+			startMessage = "<ul><li class=\"mb-parent-message\">";
+			startMessageBody = "<div class=\"mb-parent-message-body\">";
+			startMessageThread = "<div class=\"mb-parent-message-thread\">";
+			startUser = "<div class=\"mb-parent-message-user\">";
+		}
+
+		for (MBMessage mbMessage : thread) {
+			String messageUser = mbMessage.getUserName();
+
+			if (!htmlFormat) {
+				messageUser = _quotedMessage(messageUser, quoteMark);
+			}
+
+			messageParentBody = StringBundler.concat(
+				messageParentBody, startMessage, startUser, messageUser, end,
+				startMessageBody,
+				getMessageBody(
+					htmlFormat, mbMessage, quoteMark, serviceContext),
+				end, endMessage);
+
+			endElements.append(endElement);
+
+			if (!Validator.isBlank(quoteMark)) {
+				quoteMark = StringPool.GREATER_THAN + quoteMark;
```

Comment:
Instead of concatenating strings like this, why not keep a depth count? If you prefer to concatenate, please use a string bundler.

Reply:
mmmm

The part of the deep count. you are using the quoteMark inside the concat and changes every iteration.


The String Bundler:
I do not think it's a good idea, We should add the GREATER_THAN on the first position.

We should do something like

```
StringBuilder quoteMarkBuilder = new StringBuilder(thread.size()+1);

quoteMarkBuilder.append(quoteMark);

for (MBMessage mbMessage : thread) {
	messageParentBody.append(...quoteMarkBuilder.toString()...)
			...


	if (!Validator.isBlank(quoteMarkBuilder.toString())) {
		quoteMarkBuilder.insert(0, StringPool.GREATER_THAN);
	}
}
```
and the `insert` is ... meh.


Maybe, I'm missing something?